Extend geolocation tests

sadym-chromium opened the pull request:
Copy Chromium's wpt-internal test to public WPT using testdriver geolocation API.

* [wpt_internal/geolocation-api/success.https.html](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/wpt_internal/geolocation-api/success.https.html;drc=0fb74337a38fda70931714003cb8732957a69de8)
* [wpt_internal/geolocation-api/error.https.html](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/wpt_internal/geolocation-api/error.https.html)

sadym-chromium committed:
Extend geolocation tests

marcoscaceres commented on geolocation/enabled.https.html:
```diff
@@ -0,0 +1,35 @@
+            {coordinates: null});
+    });
+
+    promise_test(async (t) => {
```
This test seems unrelated to the above. It's also tested by the IDL harness tests, so we probably don't need this.

sadym-chromium replied:
Agree, removed.

marcoscaceres commented on geolocation/enabled.https.html:
```diff
@@ -0,0 +1,35 @@
+        });
+
+        // Ensure any previously set geolocation emulations are cleared.
+        await test_driver.bidi.emulation.set_geolocation_override(
```
I don't think we have agreement on the shape of the BiDi API? @reillyeon?

sadym-chromium replied:
The API is specified here:
* https://www.w3.org/TR/geolocation/#emulation
* https://www.w3.org/TR/webdriver-bidi/#command-emulation-setGeolocationOverride

reillyeon replied:
https://github.com/w3c/geolocation/pull/183 was merged based on consensus between Chromium and Firefox.

marcoscaceres commented on geolocation/success.https.html:
```diff
@@ -0,0 +1,46 @@
+        const mockLongitude = -0.166;
+        const mockAccuracy = 100;
+
+        let position;
```
```suggestion
```

sadym-chromium replied:
done

marcoscaceres commented on geolocation/success.https.html:
```diff
@@ -0,0 +1,46 @@
+                }
+            });
+
+        navigator.geolocation.getCurrentPosition(function (p) {
+            position = p;
```
```suggestion
        navigator.geolocation.getCurrentPosition((position) => {
```

marcoscaceres commented on geolocation/success.https.html:
```diff
@@ -0,0 +1,46 @@
+
+        let position;
+
+        await test_driver.bidi.emulation.set_geolocation_override(
```
This seems potentially racy. it might be better to hold promises for `navigator.geolocation.getCurrentPosition((position)` and `test_driver.bidi.emulation.set_geolocation_override`, and then await each, and then check the position result from navigator.geolocation.getCurrentPosition(), like:

```
const setPositionPromise = test_driver.bidi.emulation.set_geolocation_override(...);
const getPositionPromise = new Promise((resolve, reject) => navigator.geolocation.getCurrentPosition(resolve, reject);
await setPositionPromise;
try {
   const position = await getPositionPromise;
    // assert test things here 
} catch (err) {
   assert_unreached
}
const position 
```

sadym-chromium replied:
Agree. Fixed.

marcoscaceres reviewed (dismissed):
Some of these things are a bit redundant or not quite correct and potentially racy.

sadym-chromium committed:
extract `getCurrentGeolocation` to support.js

sadym-chromium committed:
refactor

reillyeon commented on geolocation/success.https.html:
```diff
@@ -0,0 +1,43 @@
+        const longitude = -0.166;
+        const accuracy = 100;
+
+        await test_driver.bidi.emulation.set_geolocation_override({
```
Use `t.add_cleanup()` to reset this override after the test completes.

reillyeon commented on geolocation/success.https.html:
```diff
@@ -0,0 +1,43 @@
+            coordinates: {latitude, longitude, accuracy}
+        });
+
+        const positionPromise = new Promise(
+            (resolve, reject) =>
+                window.navigator.geolocation.getCurrentPosition(
+                    position => resolve(position.coords.toJSON()),
+                    error => reject(error),
+                    {timeout: 200}
+                ))
+        const position = await positionPromise;
```
```suggestion
        const position = await new Promise(
            (resolve, reject) =>
                window.navigator.geolocation.getCurrentPosition(
                    position => resolve(position.coords.toJSON()),
                    error => reject(error),
                    {timeout: 200}
                ));
```

sadym-chromium committed:
rename test file

sadym-chromium commented:
@marcoscaceres please take a look

sadym-chromium committed:
Test `POSITION_UNAVAILABLE`

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/wpt_internal/geolocation-api/error.https.html

sadym-chromium commented:
I assume the concerns raised by @marcoscaceres are addressed
